[don't merge] Copies of Approach of the Second Sun now cause lifegain

the-red-lily opened the pull request:
Resolves #9126
Following suggestions from #9156
Copies of Approach of the Second Sun now cause the controller to gain life

the-red-lily commented:
@JayDi85 On the previous PR you mentioned writing test cases for this. Do you want me to take a crack at it? I don't want to double up work if you've already done it though

Alex-Vasile commented:
For future reference, you don't have to close the PR to add changes. Simply make more commits to the same branch and push them. They will update the PR

the-red-lily commented:
> For future reference, you don't have to close the PR to add changes. Simply make more commits to the same branch and push them. They will update the PR

Are force-pushes okay? I wanted to rewrite history

the-red-lily commented:
> @the-red-lily original "fix" was in additional null check only. So if it was a workable fix then the original bug was in empty card field in AI games. So my unit test was for AI related games and potential bug (unproduceable yet).

I'm not sure what AI games you are discussing. The original issue #9126 only references [[Fork]] copying [[Approach of the Second Sun]] not causing lifegain. There was discussion from my end, where I got confused during debugging due to AI simulation of the game, but that was not the cause of the bug.
> > A copy of a spell isn’t cast, so it won’t count as the first nor as the second Approach of the Second Sun. (2017-04-18)
> 
> BTW you must add that rules ref as comments to the code
Will do

the-red-lily commented:
AFAIK I've addressed all comments and added tests. The tests cover the bug here, and all rulings found on Gatherer. Let me know if it needs anything further before a merge

the-red-lily commented:
@JayDi85 is anything further needed before merge?

Alex-Vasile commented on Mage.Sets/src/mage/cards/a/ApproachOfTheSecondSun.java:
```diff
@@ -64,28 +64,30 @@ public ApproachOfTheSecondSunEffect copy() {
+        }
+        //If this spell was cast from your hand and you've cast another spell named {this} this game
+        //A copy of a spell isn’t cast, so it won’t count as the first nor as the second Approach of the Second Sun. (2017-04-18)
+        if (!spell.isCopy() //TODO: copied spells should not be "from" hand
```
What's up with this TODO?

the-red-lily replied:
The point of my comment here, is that in my view, ApproachOfTheSecondSun should not be manually checking for copies. I think checking the "fromZone" field should be sufficient. But that would require a bit of a deeper dive into spell copying and zone changes, it would need a new PR and more knowledge of MTG rules plus xmage engine

Alex-Vasile commented on Mage.Sets/src/mage/cards/a/ApproachOfTheSecondSun.java:
```diff
@@ -64,28 +64,30 @@ public ApproachOfTheSecondSunEffect copy() {
+        if (controller == null || spell == null || watcher == null) {
+            return false;
+        }
+        //If this spell was cast from your hand and you've cast another spell named {this} this game
```
Nitpicky, but can you please make sure to leave the space between the "//" and the comment.
